modules/statistics/CallStats.js

```diff
@@ -68,164 +70,330 @@ const reportType = {
+            // download did not succeed in general or on time). Further attempts
+            // to utilize it cannot possibly succeed.
             GlobalOnErrorHandler.callErrorHandler(e);
+            backend = null;
             logger.error(e);
-        }
-    };
-}
 
-/* eslint-disable no-invalid-this */
-
-/**
- * Creates new CallStats instance that handles all callstats API calls.
- * @param peerConnection {JingleSessionPC} the session object
- * @param options {object} credentials for callstats.
- */
-const CallStats = tryCatch(function(jingleSession, options) {
-    try {
-        CallStats.feedbackEnabled = false;
-        callStats = new callstats($, io, jsSHA); // eslint-disable-line new-cap
+            return false;
+        }
+    }
 
-        this.peerconnection = jingleSession.peerconnection.peerconnection;
+    /* eslint-disable no-invalid-this */
 
-        this.userID = {
-            aliasName: Strophe.getResourceFromJid(jingleSession.room.myroomjid),
-            userName: Settings.getCallStatsUserName()
-        };
+    /**
+     *
+     * @param err
+     * @param msg
+     */
+    static _initCallback(err, msg) {
+        logger.log(`CallStats Status: err=${err} msg=${msg}`);
 
-        // The confID is case sensitive!!!
-        this.confID = `${options.callStatsConfIDNamespace}/${options.roomName}`;
+        // there is no lib, nothing to report to
+        if (err !== 'success') {
+            return;
+        }
 
-        this.callStatsID = options.callStatsID;
-        this.callStatsSecret = options.callStatsSecret;
+        // I hate that
+        let atLeastOneFabric = false;
+        let defaultInstance = null;
+
+        for (const callStatsInstance of CallStats.fabrics.values()) {
+            if (!callStatsInstance.hasFabric) {
+                logger.debug('addNewFabric - initCallback');
+                if (callStatsInstance._addNewFabric()) {
+                    atLeastOneFabric = true;
+                    if (!defaultInstance) {
+                        defaultInstance = callStatsInstance;
+                    }
+                }
+            }
+        }
 
-        CallStats.initializeInProgress = true;
+        if (!atLeastOneFabric) {
+            return;
+        }
 
-        // userID is generated or given by the origin server
-        callStats.initialize(this.callStatsID,
-            this.callStatsSecret,
-            this.userID,
-            initCallback.bind(this));
+        CallStats.initialized = true;
+
+        // There is no conference ID nor a PeerConnection available when some of
+        // the events are scheduled on the reportsQueue, so those will be
+        // reported on the first initialized fabric.
+        const defaultConfID = defaultInstance.confID;
+        const defaultPC = defaultInstance.peerconnection;
+
+        // notify callstats about failures if there were any
+        if (CallStats.reportsQueue.length) {
```

Comment:
You can use a for..of and get rid of the extra level of indentation (no need to check for the length).